File: app/src/main/java/com/example/shoppinglist/repositories/ProductRepository.java
package com.example.shoppinglist.repositories;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import com.example.shoppinglist.database.ProductDao;
import com.example.shoppinglist.database.ShoppingListDatabase;
import com.example.shoppinglist.models.Product;

import java.util.List;

public class ProductRepository {

    private ProductDao productDao;
    private LiveData<List<Product>> allProduct;

    public ProductRepository(Application application, int shoppingListId) {
        ShoppingListDatabase database = ShoppingListDatabase.getInstance(application);
        productDao = database.productDao();
        allProduct = productDao.getAllProduct(shoppingListId);
    }

    public void insertProduct(Product product) {
        new InsertProductAsyncTask(productDao).execute(product);
    }

    public void updateProduct(Product product) {
        new UpdateProductAsyncTask(productDao).execute(product);
    }

    public void deleteProduct(Product product) {
        new DeleteProductAsyncTask(productDao).execute(product);
    }

    public LiveData<List<Product>> getAllProduct() {
        return allProduct;
    }

    private static class InsertProductAsyncTask extends AsyncTask<Product, Void, Void> {
        private ProductDao productDao;

        private InsertProductAsyncTask(ProductDao productDao) {
            this.productDao = productDao;
        }

        @Override
        protected Void doInBackground(Product... products) {
            productDao.insertProduct(products[0]);
            return null;
        }
    }

    private static class UpdateProductAsyncTask extends AsyncTask<Product, Void, Void> {
        private ProductDao productDao;

        private UpdateProductAsyncTask(ProductDao productDao) {
            this.productDao = productDao;
        }

        @Override
        protected Void doInBackground(Product... products) {
            productDao.updateProduct(products[0]);
            return null;
        }
    }

    private static class DeleteProductAsyncTask extends AsyncTask<Product, Void, Void> {
        private ProductDao productDao;

        private DeleteProductAsyncTask(ProductDao productDao) {
            this.productDao = productDao;
        }

        @Override
        protected Void doInBackground(Product... products) {
            productDao.deleteProduct(products[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/example/shoppinglist/repositories/ProductRepository.kt
package com.example.myapplication22

import android.app.Application
import androidx.lifecycle.LiveData
import com.example.shoppinglist.database.ProductDao
import com.example.shoppinglist.database.ShoppingListDatabase
import com.example.shoppinglist.models.Product
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.launch

class ProductRepository(application: Application?) {
    private val productDao: ProductDao

    fun insertProduct(product: Product) {
        CoroutineScope(IO).launch {
            insertProductOnBg(productDao, product)
        }
    }

    fun updateProduct(product: Product){
        CoroutineScope(IO).launch {
            updateProductOnBg(productDao, product)
        }
    }

    fun deleteProduct(product: Product){
        CoroutineScope(IO).launch {
            deleteProductOnBg(productDao, product)
        }
    }

    fun getAllProduct(shoppingListId: Int): LiveData<List<Product>> {
        return productDao.getAllProduct(shoppingListId)
    }

    init {
        val database = ShoppingListDatabase.getInstance(application)
        productDao = database.productDao()
    }

    private fun insertProductOnBg(dao: ProductDao, product: Product) {
        dao.insertProduct(product)
    }

    private fun updateProductOnBg(dao: ProductDao, product: Product) {
        dao.updateProduct(product)
    }

    private fun deleteProductOnBg(dao: ProductDao, product: Product) {
        dao.deleteProduct(product)
    }
}
